# Tips Data Set - Project 2019
### Programing and Scripting
### GMIT Galway
#### Slawomir Sowa 
#### 04/10/2019

## Objectives

* Create a git repository
* discusses and analyses whether there is a relationship between the total bill and tip amount,
* analyse the relationship between the variables within the dataset.

Full description of Assignment can be found <a href="">here</a>

### Tools
To analyse Iris Data Set I work with Python 3.7 and Jupyter Notebook.

<img width='400px' align='left' src='https://www.python.org/static/community_logos/python-logo-master-v3-TM.png'>

<img width='150px' align='left' src='https://jupyter.org/assets/main-logo.svg'>



#### Used Libraries and tools
#### Pandas
Pandas is a Python package providing fast, flexible, and expressive data structures designed to make working with “relational” or “labeled” data both easy and intuitive. The two primary data structures of pandas, Series (1-dimensional) and DataFrame (2-dimensional), handle the vast majority of typical use cases in finance, statistics, social science, and many areas of engineering. http://pandas.pydata.org

#### NumPy
NumPy is the fundamental package for scientific computing with Python. It contains among other things: a powerful N-dimensional array object, sophisticated (broadcasting) functions, useful linear algebra, Fourier transform, and random number capabilities, Besides its obvious scientific uses, NumPy can also be used as an efficient multi-dimensional container of generic data. Arbitrary data-types can be defined. This allows NumPy to seamlessly and speedily integrate with a wide variety of databases. https://www.numpy.org/

#### Matplotlib
Matplotlib is a Python 2D plotting library which produces publication quality figures in a variety of hardcopy formats and interactive environments across platforms. Matplotlib tries to make easy things easy and hard things possible. You can generate plots, histograms, power spectra, bar charts, errorcharts, scatterplots, etc., with just a few lines of code. For examples, see the sample plots and thumbnail gallery. https://matplotlib.org/index.html

#### Seaborn
Seaborn is a library for making statistical graphics in Python. It is built on top of matplotlib and closely integrated with pandas data structures. https://seaborn.pydata.org/

#### Jupyter Notebook
The Jupyter Notebook is an open-source web application that allows to create and share documents that contain live code, equations, visualizations and narrative text. Uses include: data cleaning and transformation, numerical simulation, statistical modeling, data visualization, machine learning, and much more. https://jupyter.org/index.html

 ## 1. Setting the research goal

The main goal of the project is to analyze the relationship between the bill and the tip and check if such a relationship exists. The secondary goal is to analyze the other variables in the data set. such as customer sex, day of the week, or time of day. 

## 2. Retrieve data

In this step we collect data, ensure that we can use the data in our program, which means checking the existence of, quality, and access to data.

Our data are stored in csv file and we can fined dataset here: [tips dataset](https://github.com/Slawak1/GMIT-FDA_Assesment_2019/blob/master/tips.csv)

To work with data we need to import all necessery tools as pandas and numpy for data manipulation, matplotlib and seaborn for data visualisation. 

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns 

To read data from Tips.csv file I used pandas library and read_csv function. Variable df is my data frame. 

In [2]:
df = pd.read_csv('tips.csv') 
print(df.shape)

(244, 7)


Pandas <code>shape() </code> function returns tuple representing the dimensionality of the DataFrame. Our dataset contains 244 rows and 7 columns.

<code>df.head() </code>function returns first 5 rows from data freame. It is useful for quickly testing if your object has the right type of data in it.

In [3]:
df.head(5)

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

<code>df.tail()</code> returns last 5 rows from data frame. We can use it for veryfying data.

In [4]:
df.tail(5)

total_bill   tip     sex smoker   day    time  size
239       29.03  5.92    Male     No   Sat  Dinner     3
240       27.18  2.00  Female    Yes   Sat  Dinner     2
241       22.67  2.00    Male    Yes   Sat  Dinner     2
242       17.82  1.75    Male     No   Sat  Dinner     2
243       18.78  3.00  Female     No  Thur  Dinner     2

## 3. Data Preparation

This phase consist of three subphases:<br> <i> data cleansing </i> removes false values from a data source and inconsistencies acreoo data source, <br> <i> data integration</i> enrich data source by combining information from multiple data source and <br><i> data transformation</i> ensures that the data is in a suitable format for use.

To check is any data missing in Tips data set I used <code>isnull()</code> function from pandas library. Function returns boolean value <code>True</code> if null value exists and <code>False</code> if not null value founded. Pandas will recognise empty cell or <code>NA</code> as missing values.


In [5]:
df.isnull().any()


total_bill    False
tip           False
sex           False
smoker        False
day           False
time          False
size          False
dtype: bool

The above code shows us that there is no <code>null</code> values in our data frame.

But sometimes it might be the case where there’s missing values that have different formats than empty cell or N/A.
For example it can be '--', 'na' or 'n/a'. <br>
To find that kind of missing values we can use python code and iterate over a list of missing values.  

In [6]:
# create a list of missing values
missing = ['--','-','na','n/a'] 

# test if values exist in our data frame
for i in missing:
    if i in df.values:
        print ('Missing values exist')
    else:
        print ('No missing values detected')

No missing values detected
No missing values detected
No missing values detected
No missing values detected


In [ ]:
df.describe()

## References

Pandas Data Frame

https://pandas.pydata.org/pandas-docs/stable/reference/frame.html


"Introducing Data Science" Davy Cielen, Arno D. B. Meysman, Mohamed Ali - Manning 2016


missing values:
https://towardsdatascience.com/data-cleaning-with-python-and-pandas-detecting-missing-values-3e9c6ebcf78b